File: my-folder/0039-combination-sum/solution.go
func combinationSum(candidates []int, target int) [][]int {
    res := [][]int{}

    var dfs func(i int, curr []int, total int)
    dfs = func(i int, curr []int, total int) {
        if total == target {
            tmp := make([]int, len(curr))
            copy(tmp, curr)
            res = append(res, tmp)
            return
        }

        if i >= len(candidates) || total > target {
            return
        }

        curr = append(curr, candidates[i])
        dfs(i, curr, total+candidates[i])

        curr = curr[:len(curr)-1]
        dfs(i+1, curr, total)
    }

    dfs(0, []int{}, 0)
    return res
}

File: my-folder/0039-combination-sum/solution.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let total = 0
    let currConbination = []
    let result = []

    function backtrack(i, currConbination, total) {
        if (total > target || i >= candidates.length) {
            return
        }

        if (total === target) {
            result.push([...currConbination])
            return
        }


        currConbination.push(candidates[i])
        backtrack(i, currConbination, total + candidates[i])


        currConbination.pop()
        backtrack(i+1, currConbination, total)
    }

    backtrack(0, currConbination, total)

    return result
};
